aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -628,28 +626,40 @@ impl AptosVM {
             MoveValue::Address(proposer),
             MoveValue::U64(timestamp),
         ]);
-        session
-            .execute_function_bypass_visibility(
-                &BLOCK_MODULE,
-                BLOCK_PROLOGUE,
-                vec![],
-                args,
-                &mut gas_status,
-            )
-            .map(|_return_vals| ())
-            .or_else(|e| {
-                expect_only_successful_execution(e, BLOCK_PROLOGUE.as_str(), log_context)
-            })?;
+        let result = match session.execute_function_bypass_visibility(
+            &BLOCK_MODULE,
+            BLOCK_PROLOGUE,
+            vec![],
+            args,
+            &mut gas_status,
+        ) {
+            Ok(_) => {
+                let output = get_transaction_output(
+                    &mut (),
+                    session,
+                    gas_status.remaining_gas(),
+                    &txn_data,
+                    ExecutionStatus::Success,
+                )?;
+                (VMStatus::Executed, output)
+            }
+            Err(e) => {
+                // make sure the error is propagated back as transaction error
+                let vm_status = e.into_vm_status();
+                let transaction_status = TransactionStatus::from(vm_status.clone());
+                let output = match transaction_status {
+                    TransactionStatus::Keep(_) => {
+                        TransactionOutput::new(WriteSet::default(), vec![], 0, transaction_status)
+                    }
+                    TransactionStatus::Discard(error_code) => discard_error_output(error_code),
```

Comment:
is this effectively the change in question?
I have to say that I am not sure I understand the problem and the fix any longer.
Could you try to explain what this is going to fix?
I guess I am looking at the block prologue and the asserts look very system-wide. I mean they look like something was really wrong. I do not understand what parallel execution is doing to make those assert kick in.
I have few more comments but it would help if you could explain again (sorry!), happy to chat face-to-face if needed.

Also is this possibly covering too much? I guess some internal error is a bad thing by definition and will be caught again? In other words if things get re-executed by the parallel executor (assuming that is the intent) is it possible we re-execute forever if some subtle nasty error is there in the VM? maybe something that required the VM to restart but here we go over and over again? (I guess that is not an issue but it would help to hear from you instead of exploring the code base)

Reply:
we probably don't necessarily need this PR if we fix parallel execution properly with https://github.com/aptos-labs/aptos-core/pull/1303

the problem is that with 3 transactions tx1, tx2, tx3. If executed sequentially, they all succeed, but if we execute speculatively tx1 then tx3 without tx2, tx3 would fail and return a VMError. The bug in parallel execution uses the speculative error as confirmed result which was fixed in the PR